clippy_lints/src/extra_unused_type_parameters.rs

```diff
@@ -0,0 +1,175 @@
+    /// requires using a turbofish, which serves no purpose but to satisfy the compiler.
+    ///
+    /// ### Example
+    /// ```rust
+    /// // unused type parameters
+    /// fn unused_ty<T>(x: u8) {
+    ///     // ..
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn no_unused_ty(x: u8) {
+    ///     // ..
+    /// }
+    /// ```
+    #[clippy::version = "1.68.0"]
+    pub EXTRA_UNUSED_TYPE_PARAMETERS,
+    complexity,
+    "unused type parameters in function definitions"
+}
+declare_lint_pass!(ExtraUnusedTypeParameters => [EXTRA_UNUSED_TYPE_PARAMETERS]);
+
+struct TypeWalker<'cx, 'tcx> {
+    cx: &'cx LateContext<'tcx>,
+    map: FxHashMap<DefId, Span>,
+    bound_map: FxHashMap<DefId, Span>,
+    generics: &'tcx Generics<'tcx>,
+    some_params_used: bool,
+    has_non_ty_params: bool,
+}
+
+impl<'cx, 'tcx> TypeWalker<'cx, 'tcx> {
+    fn new(cx: &'cx LateContext<'tcx>, generics: &'tcx Generics<'tcx>) -> Self {
+        let mut has_non_ty_params = false;
+        let map = generics
+            .params
+            .iter()
+            .filter_map(|param| match &param.kind {
+                GenericParamKind::Type { .. } => Some((param.def_id.into(), param.span)),
+                GenericParamKind::Lifetime {
+                    kind: LifetimeParamKind::Elided,
+                } => None,
+                _ => {
+                    has_non_ty_params = true;
+                    None
+                },
+            })
+            .collect();
+        Self {
+            cx,
+            map,
+            generics,
+            has_non_ty_params,
+            bound_map: FxHashMap::default(),
+            some_params_used: false,
+        }
+    }
+
+    fn emit_lint(&self) {
+        let (msg, help) = match self.map.len() {
+            0 => return,
+            1 => (
+                "type parameter goes unused in function definition",
+                "consider removing the parameter",
+            ),
+            _ => (
+                "type parameters go unused in function definition",
+                "consider removing the parameters",
+            ),
+        };
+
+        let source_map = self.cx.tcx.sess.source_map();
+        let span = if self.some_params_used || self.has_non_ty_params {
+            MultiSpan::from_spans(
+                self.map
+                    .iter()
+                    .map(|(def_id, &(mut span))| {
+                        if let Some(bound_span) = self.bound_map.get(def_id) {
+                            span = span.with_hi(bound_span.hi());
+                        }
+                        span = source_map.span_extend_to_next_char(span, ',', false);
+                        span = span.with_hi(BytePos(span.hi().0 + 1));
+
+                        let max_hi = self.generics.span.hi();
+                        if span.hi() >= max_hi {
+                            span = span.with_hi(BytePos(max_hi.0 - 1));
```

Comment:
The first use there is to include the comma after the parameter when highlighting the span. Since `SourceMap::span_extend_to_next_char` stops at the character immediately prior, advancing one character farther should always work, I think.

The second use occurs if the comma scanning blows past the end of the list of parameters (and into the list of function arguments), and rewinds back to the character just before the last ">".

Reply:
Makes sense, although the presence of any bounds that are themselves generic would complicate this, since they contain `>`, and might also contain commas.

Reply:
I see. Yeah, then I think your solution is the best. But please add comments why those `BytePos`es are "safe".

Reply:
I think it's still possible to get rid of them actually. I'll just have to skip past the bound if it's present, then apply the logic you posted.